### PR TITLE
Add LLaMA-2 support

### PR DESCRIPTION
**Pull request by prusnak**
Followup to https://github.com/tinygrad/tinygrad/pull/1275 which cleans up code reuse and introduces LLaMA-2 support.

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/state.py                          1               1
------------------------------------------------------------
lines added in the tinygrad folder: 0
```

File: tinygrad/state.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def _rebuild_tensor_v2(storage, storage_offset, size, stride, requires_grad, bac
 
     return ret.reshape(size)
 
-  intercept = {"HalfStorage": dtypes.float16, "FloatStorage": dtypes.float32, "IntStorage": dtypes.int32, "LongStorage": dtypes.int64, "_rebuild_tensor_v2": _rebuild_tensor_v2}
+  intercept = {"HalfStorage": dtypes.float16, "FloatStorage": dtypes.float32, "BFloat16Storage": dtypes.float16, "IntStorage": dtypes.int32, "LongStorage": dtypes.int64, "_rebuild_tensor_v2": _rebuild_tensor_v2}
```
Comment (geohot) [inline tinygrad/state.py:83]: Err, this isn't right. dtypes.float16 isn't bfloat

Comment (prusnak) [reply tinygrad/state.py:83]: Right. I was trying to find a way how to hack bfloat16 support into tinygrad, but I failed.

Simply adding a new dtype to `helpers.dtype` did not do the trick, because the backend does not know this.

Please advise, how to proceed.

It seems the correct way would be to use float32 as storage and use code similar to this one to convert bfloat16 to float32:

https://github.com/ggerganov/llama.cpp/blob/294f424554c1599784ac9962462fc39ace92d8a5/convert.py#L338-L341

Comment (dc-dc-dc) [inline tinygrad/state.py:83]: bfloat16 is 8bits exponent 7 bits fraction while float16 is 5bits exponent 10 bits fraction. Can you verify that this is not causing a accuracy loss in the weight values?

Comment (prusnak) [reply tinygrad/state.py:83]: It does not work at all unfortunately. Please check my comment above: https://github.com/tinygrad/tinygrad/pull/1284#discussion_r1268666825

**Comment by geohot** [comment]
Do they all need bfloat16? Will merge without if they don't need it.

**Comment by prusnak** [comment]
> Do they all need bfloat16? Will merge without if they don't need it.

The smallest one needs it, so I assume bigger ones need it too.

File: tinygrad/state.py (hunk 1 above)
Comment (dc-dc-dc) [reply tinygrad/state.py:83]: Did a quick test locally and it matched up with pytorch but was pretty slow and double the memory footprint

```python
    if storage[1] == dtypes._bfloat16:
      temp = t[byte_offset:byte_offset+prod(size)].cast(dtypes.uint16)
      ret = Tensor((temp.numpy().astype(np.uint32) << 16).view(np.float32))
    else:
      ret = t[byte_offset:byte_offset+prod(size)].cast(storage[1])
```

Comment (prusnak) [reply tinygrad/state.py:83]: @dc-dc-dc I think that's better than nothing, we can always optimize later. Will you please push your changes into my PR branch? If that's not possible, will you please share your changes with me so I can update the PR?

**Comment by geohot** [comment]
I think bfloat16 should be added in a separate PR. It's okay if only some backends support it.

File: tinygrad/state.py (hunk 1 above)
Comment (dc-dc-dc) [reply tinygrad/state.py:83]: Not able to make a pr to your fork for some reason but I pushed the changes [here](https://github.com/dc-dc-dc/tinygrad/tree/llama-refactor)

**Comment by prusnak** [comment]
I see the PR https://github.com/tinygrad/tinygrad/pull/1293 has been merged

Rebased on top of the current master and change commit https://github.com/tinygrad/tinygrad/pull/1284/commits/cfc227edb86e452b892a78bf989b95ea2c1fd0c5 to use `dtypes.bfloat16` for `BFloat16Storage`

**Comment by prusnak** [comment]
Hmmm ... https://github.com/tinygrad/tinygrad/issues/1290#issuecomment-1642880556 suggets we should transform bf16->f32->f16 because the upstream LLaMA inference scripts use f16 anyway and therefore perform this conversion.

LMK how should we proceed with this PR.

**Comment by prusnak** [comment]
I just wanted to know which way to proceed is preferred:
- bfloat16 implementation that works only on llvm
- converting bfloat16 to float16 which works everywhere

Since it seems that LLaMA upstream inference code does the latter, I will proceed with option 2 and update the PR to have working implementation.

**Comment by prusnak** [comment]
I updated the PR with commit 5a6f288c71b0e12dc3cd3125308cde0e169dc22f which converts bfloat16->float32->float16 like suggested by @sekstini in https://github.com/tinygrad/tinygrad/issues/1290#issuecomment-1642880556

I confirm that with this commit the LLaMA-2 (`python3 examples/llama.py --gen 2`) works for me (M1) both with and without `MPS=1` set:

File: tinygrad/helpers.py [hunk 2]
```diff
@@ -96,16 +96,19 @@ def fields() -> Dict[str, DType]: return DTYPES_DICT
...
-  uint64: Final[DType] = DType(2, 8, "unsigned long", np.uint64)
+  uint16: Final[DType] = DType(1, 2, "unsigned short", np.uint16)
+  uint32: Final[DType] = DType(2, 4, "unsigned int", np.uint32)
+  uint64: Final[DType] = DType(3, 8, "unsigned long", np.uint64)
 
   # NOTE: bfloat16 isn't supported in numpy
   bfloat16: Final[DType] = DType(0, 2, "__bf16", None)
 
   # NOTE: these are internal dtypes, should probably check for that
+  _bfloat16: Final[DType] = DType(0, 2, "bfloat16", None)
```
Comment (wozeparrot) [inline tinygrad/helpers.py:111]: Why can't the `bfloat16` dtype above be used?

Comment (prusnak) [reply tinygrad/helpers.py:111]: Because then it works only on LLVM and no other backend.

Comment (wozeparrot) [reply tinygrad/helpers.py:111]: The idea with it was to load the weights to LLVM cast to fp16, then move them to whatever device they will be used on.

Comment (prusnak) [reply tinygrad/helpers.py:111]: That sounds good. However, I have no idea how to do this because of my superficial knowledge of the tinygrad architecture. Feel free to take over this PR and update it to match your expectations.

Comment (sekstini) [reply tinygrad/helpers.py:111]: If we're using pytorch for the loading anyway, why not just do `.half()` after loading?
(I confirmed `x.half()` gave the same result as `x.float().half()`)

Comment (sekstini) [reply tinygrad/helpers.py:111]: Ah nevermind, I see we're using the tinygrad torch loader. My bad.

Comment (prusnak) [reply tinygrad/helpers.py:111]: > The idea with it was to load the weights to LLVM cast to fp16

@wozeparrot can you provide some hints how to do this so I can rework the PR?

Comment (wozeparrot) [reply tinygrad/helpers.py:111]: You should be able to do something like: `x.to("llvm").float().realize().to(Device.DEFAULT).realize()` where `x` is the loaded weight

Comment (prusnak) [reply tinygrad/helpers.py:111]: Hm, I tried the following:

``` python
    if storage[1] == dtypes.bfloat16:
      ret = t[byte_offset:byte_offset+prod(size)].to("LLVM").half().realize().to(t.device).realize()
    else:
      ret = t[byte_offset:byte_offset+prod(size)].cast(storage[1])
```

and it seems it modified weights on my disk :-/ Redownloading LLaMA-2-7b weights again ...

Comment (wozeparrot) [reply tinygrad/helpers.py:111]: oh yea can't do t.device because the weights are mmapped from disk, so the tensor's device is disk. So moving the weights to that device is the same as writing to disk.

Comment (prusnak) [reply tinygrad/helpers.py:111]: > so the tensor's device is disk

Learned the hard way, but it makes sense! :-)

Comment (prusnak) [reply tinygrad/helpers.py]: Pushed bf167b8db1fd5f3e5a2125b1d35f0abd9262ba0c which converts BFloat16 to Float16 via LLVM, but I am unable to get meaningful results (Stacy keeps talking jibberish).

I am once again out of ideas.

Comment (wozeparrot) [reply tinygrad/helpers.py]: maybe try something like:

`ret = t[byte_offset:byte_offset+prod(size)].contiguous().realize().to("LLVM").realize().half().realize().to(Device.DEFAULT).realize()`

Comment (prusnak) [reply tinygrad/helpers.py]: @wozeparrot nope, that didn't help

**Comment by prusnak** [comment]
After the debugging session with @wozeparrot on Discord, we came up with the following commit:

https://github.com/tinygrad/tinygrad/pull/1284/commits/b42cddb410843e8a00a0a3fc906a55a328211eb4

Everything now works as expected (Stacy responds with meaningful answers), and we convert `bfloat16` to `float16` without the need to use NumPy.

File: tinygrad/state.py [hunk 3]
```diff
@@ -67,6 +68,11 @@ def _rebuild_tensor_v2(storage, storage_offset, size, stride, requires_grad, bac
     if storage[2] not in offsets: return None
     byte_offset = offsets[storage[2]]+storage_offset*storage[1].itemsize
     ret = t[byte_offset:byte_offset+prod(size)].cast(storage[1])
+    # convert bfloat16 -> float16 using LLVM
+    # upstream LLaMA also does this conversion:
```
Comment (wozeparrot) [inline tinygrad/state.py:72]: So the one thing is that other model do use this, but I think its fine that we always convert bfloat16 to float16 for now, as we don't have great bfloat16 support.
